internal/cmds/cmds.go

```diff
@@ -370,6 +390,110 @@ func CopyStateForUpdate(ctx log.Logger) error {
 	return nil
 }
 
+func rehydrateMrSeqFilesForProblematicUpgrades(ctx *log.Context, h types.HandlerEnvironment) error {
+	// First, determine whether we're upgrading from a 'problematic' version, defined as one
+	// where we mistakenly deleted the mrseq files in the Disable call
+	newExtensionVersion := os.Getenv(constants.ExtensionVersionEnvName)
+	oldExtensionVersion := os.Getenv(constants.ExtensionVersionUpdatingFromEnvName)
+	newExtensionDirectory := os.Getenv(constants.ExtensionPathEnvName)
+	oldExtensionDirectory := strings.ReplaceAll(newExtensionDirectory, newExtensionVersion, oldExtensionVersion)
+
+	// The following are problematic versions:
+	// Production: 1.3.17
+	// Test: 1.8.0, 1.9.0
+	isProblematicVersion := false
+	isTestExtension := strings.Contains(oldExtensionDirectory, "Microsoft.Azure.Extensions.Edp.RunCommandHandlerLinuxTest")
+	if isTestExtension {
+		isProblematicVersion = (oldExtensionVersion == "1.8.0" || oldExtensionVersion == "1.9.0")
+	} else {
+		isProblematicVersion = (oldExtensionVersion == "1.3.17")
+	}
+
+	if isProblematicVersion {
+		ctx.Log("message", "Rehydrating mrseq files from version '%s'", oldExtensionVersion)
+		return doRehydrateMrSeqFilesForProblematicUpgrades(ctx, oldExtensionDirectory, newExtensionDirectory)
+	} else {
+		ctx.Log("message", "Previous extension version '%s' does not require mrseq hydration", oldExtensionVersion)
+	}
+
+	return nil
+}
+
+func doRehydrateMrSeqFilesForProblematicUpgrades(ctx *log.Context, oldExtensionDirectory string, newExtensionDirectory string) error {
+	oldExtensionStatusDirectory := filepath.Join(oldExtensionDirectory, constants.StatusFileDirectory)
+
+	extensionStatusDirectoryFDRef, err := os.Open(oldExtensionStatusDirectory)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to open status directory '%s'", oldExtensionStatusDirectory))
+	}
+
+	// Iterate through the status directory, looking for files of the format: {extension name}.{seqNo}.status
+	// For each of these, look for the corresponding file {extension name}.mrseq under the extension directory
+	// If we find any status files missing their corresponding mrseq, then rehydrate it by taking the seqNo from the status file name
+	statusFiles, err := extensionStatusDirectoryFDRef.ReadDir(0)
```

Comment:
is every sequence number (that is delivered to the extension) guaranteed to have a corresponding status file associated with it? afaik a seq num only has settings file if new settings were provided as part of the seq upgrade. I am just trying to make sure we don't have a blind spot or understand if we have one.

Reply:
We only need the last one, but the status file is written before the command is started, so we should have a status file.